File: Intersection_2Arrays.java
/**
 *Brute Force: check each ele in the other ele O(m*n)
 *HashMap<number, count> -- m+n
 * 2 ptrs: O(max(mlogm,nlogn))
 */
class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        int p1 = 0,p2 =0;
        List<Integer> res = new ArrayList<Integer>();
        int[] retres;
        Arrays.sort(nums1);
        Arrays.sort(nums2);
        while(p1<nums1.length && p2<nums2.length){
            if(nums1[p1] == nums2[p2]){
                res.add(nums1[p1]);
                p1++;
                p2++;
            }else if(nums1[p1] < nums2[p2])
                p1++;
            else p2++;
        }
        retres = new int[res.size()];
        for(int i=0;i<res.size();i++)
            retres[i] = res.get(i);
        return retres;
    }
}

File: intersection2arrays.py
# Time Complexity : Add - O(mlogm + n logn)
# Space Complexity :O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No
'''
1. Sorting both arrays and then using 2 pointers
2. If value is same add it to reult and move both pointer
3. If value not same move away from low value
4. When either pointer overflows, stop!
'''


from collections import defaultdict
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        
        if len(nums1) < 1 or len(nums2)<1:
            return
        
        nums1.sort()
        nums2.sort()
        
        i, j = 0, 0
        
        result = []
        while i<len(nums1) and j < len(nums2):
            
            if nums1[i] == nums2[j]:
                result.append(nums1[i])
                i += 1
                j += 1
            
            elif nums1[i] < nums2[j]:
                i += 1
                
            else:
                j += 1
        
        return result
                
# Hashmap solution   
# Time Complexity : Add - O(m)
# Space Complexity :O(n)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No
'''
1. put all values of any array(amaller or bigger) with thier frequencies in hashmap
2. Iterate thry the other array while checking the frequeency if it is > 0
3. If freq is > 0, add the value to result and decrease frequency
'''

from collections import defaultdict
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        
        if len(nums1) < 1 or len(nums2)<1:
            return
        
        if len(nums1) <= len(nums2):
            smaller = nums1
            bigger = nums2
        else:
            smaller = nums2
            bigger = nums1
           
        hash_ = defaultdict(lambda : 0)
        for i in smaller:
            hash_[i] += 1
            
        result = []
        for i in bigger:
            if i in hash_ and hash_[i] > 0:
                result.append(i)
                hash_[i] -= 1
                
        return result
                